### PR TITLE
Adding PathType to Ingress

### PR DESCRIPTION
**Pull request by cmluciano**
Co-authored-by: Christopher M. Luciano <cmluciano@us.ibm.com>

**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:
This PR implements PathType within the Ingress package and also prepares for 1.19 validation loosening of required regex path types.

**Which issue(s) this PR fixes**:
xref: 

**Special notes for your reviewer**:
- [ ] Test Create and Update to be sure regex isn't relaxed for 1.18
- [ ] Generate code properly

**Does this PR introduce a user-facing change?**:

```release-note
ingress: Add Exact and Prefix maching to Ingress PathTypes
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

[Ingress KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20190125-ingress-api-group.md)

### CODE REVIEW

**Comment by cmluciano** [comment]
cc @robscott 

No need to fully review yet, just checkpointing work

File: pkg/apis/networking/validation/validation_test.go [hunk 1]
```diff
@@ -1061,6 +1224,24 @@ func TestValidateIngressTLS(t *testing.T) {
 			}
 		}
 	}
+
```
Comment (cmluciano) [inline pkg/apis/networking/validation/validation_test.go]: Will remove this and put in a different commit since it is part of wildcard hostnames and not specific to PathType

File: staging/src/k8s.io/api/extensions/v1beta1/types.go [hunk 2]
```diff
@@ -620,18 +620,27 @@ type IngressStatus struct {
...
+	// 1. IPs are not allowed. Currently an IngressRuleValue can only apply to
+	//    the IP in the Spec of the parent Ingress.
 	// 2. The `:` delimiter is not respected because ports are not allowed.
 	//	  Currently the port of an Ingress is implicitly :80 for http and
 	//	  :443 for https.
 	// Both these may change in the future.
-	// Incoming requests are matched against the host before the IngressRuleValue.
-	// If the host is unspecified, the Ingress routes all traffic based on the
-	// specified IngressRuleValue.
+	// Incoming requests are matched against the host before the
```
Comment (cmluciano) [inline staging/src/k8s.io/api/extensions/v1beta1/types.go]: Will also remove this since it is part of wildcard host commit

File: staging/src/k8s.io/api/networking/v1beta1/types.go [hunk 3]
```diff
@@ -114,18 +114,27 @@ type IngressStatus struct {
 // the related backend services. Incoming requests are first evaluated for a host
 // match, then routed to the backend associated with the matching IngressRuleValue.
 type IngressRule struct {
-	// Host is the fully qualified domain name of a network host, as defined
-	// by RFC 3986. Note the following deviations from the "host" part of the
-	// URI as defined in the RFC:
-	// 1. IPs are not allowed. Currently an IngressRuleValue can only apply to the
-	//	  IP in the Spec of the parent Ingress.
+	// Note the following deviations from the "host" part of the
```
Comment (cmluciano) [inline staging/src/k8s.io/api/networking/v1beta1/types.go]: Remove for wildcard hostname commit

**Review by cmluciano — COMMENTED** [review]
Leaving some comments for myself on things to remove and stage to other commits

File: pkg/apis/networking/validation/validation.go [hunk 4]
```diff
@@ -315,3 +381,14 @@ func validateIngressBackend(backend *networking.IngressBackend, fldPath *field.P
 	allErrs = append(allErrs, apivalidation.ValidatePortNumOrName(backend.ServicePort, fldPath.Child("servicePort"))...)
 	return allErrs
 }
+
+// defaultBackendFieldName returns the name of the field used for defaultBackend
+// in the provided GroupVersion.
+func defaultBackendFieldName(gv schema.GroupVersion) string {
```
Comment (robscott) [inline pkg/apis/networking/validation/validation.go]: As much as I like this function, it should probably be left for a defaultBackend implementation in v1.

**Review by robscott — COMMENTED** [review]
Thanks for your work on this!

**Comment by cmluciano** [comment]
cc @liggitt

File: pkg/apis/extensions/v1beta1/defaults.go [hunk 5]
```diff
@@ -165,3 +167,9 @@ func SetDefaults_NetworkPolicy(obj *extensionsv1beta1.NetworkPolicy) {
 		}
 	}
 }
+
+func SetDefaults_HTTPIngressPath(obj *extensionsv1beta1.HTTPIngressPath) {
+	if obj.PathType == nil {
+		obj.PathType = &defaultPathType
```
Comment (liggitt) [inline pkg/apis/extensions/v1beta1/defaults.go]: Prefer a pointer to a local value. Returning a pointer to a package variable here means that any caller that assigns to that pointer will override the global default value. See https://play.golang.org/p/rGUjo2eB8dz

File: pkg/apis/networking/types.go [hunk 6]
```diff
@@ -336,19 +338,45 @@ type HTTPIngressRuleValue struct {
...
+	// PathTypeExact matches the URL path exactly.
+	PathTypeExact = PathType("Exact")
+
+	// PathTypePrefix matches based on a URL path prefix split by '/'. Matching
+	// is done on a path element by element basis. A path element refers is the
+	// list of labels in the path split by the '/' separator. A request is a
+	// match for path p if every p is an element-wise prefix of p of the request
+	// path. Note that if the last element of the path is a substring of the
+	// last element in request path, it is not a match (e.g. /foo/bar matches
+	// /foo/bar/baz, but does not match /foo/barbaz).
```
Comment (liggitt) [inline pkg/apis/networking/types.go]: clarify trailing slash behavior?
* does `path: /foo` match a request of `/foo/`?
* does `path: /foo/` match a request of `/foo`?

File: pkg/apis/networking/types.go [hunk 7]
```diff
@@ -336,19 +338,45 @@ type HTTPIngressRuleValue struct {
...
+	PathTypePrefix = PathType("Prefix")
+
+	// PathTypeImplementationSpecific matching is up to the IngressClass.
+	// Interpretation of the Path matching is up to the IngressClass.
```
Comment (liggitt) [inline pkg/apis/networking/types.go]: dedupe first two sentences

File: pkg/apis/networking/v1beta1/defaults.go [hunk 8]
```diff
@@ -17,9 +17,18 @@ limitations under the License.
...
 
+var defaultPathType = networkingv1beta1.PathTypeImplementationSpecific
+
 func addDefaultingFuncs(scheme *runtime.Scheme) error {
 	return RegisterDefaults(scheme)
 }
+
+func SetDefaults_HTTPIngressPath(obj *networkingv1beta1.HTTPIngressPath) {
+	if obj.PathType == nil {
+		obj.PathType = &defaultPathType
```
Comment (liggitt) [inline pkg/apis/networking/v1beta1/defaults.go]: same comment about not returning a pointer to a package variable

File: pkg/apis/networking/validation/validation.go [hunk 9]
```diff
@@ -174,15 +182,47 @@ func ValidateIPBlock(ipb *networking.IPBlock, fldPath *field.Path) field.ErrorLi
...
+var ValidateIngressName = apimachineryvalidation.NameIsDNSSubdomain
+
+
+// Validation options for new beta to GA transitions for HTTP PathType
+type IngressValidationOptions struct {
+	requireRegexPath bool
+}
+
+// ValidateIngress validates Ingresses on create and update.
+func ValidateIngress(ingress *networking.Ingress, opts IngressValidationOptions, requestGV schema.GroupVersion) field.ErrorList {
```
Comment (liggitt) [inline pkg/apis/networking/validation/validation.go]: suggest making ValidateIngress private so no external callers accidentally call this instead of the Create or Update-specific validation

File: pkg/apis/networking/validation/validation.go [hunk 10]
```diff
@@ -174,15 +182,47 @@ func ValidateIPBlock(ipb *networking.IPBlock, fldPath *field.Path) field.ErrorLi
...
-// ValidateIngressName validates that the given name can be used as an Ingress name.
-var ValidateIngressName = apimachineryvalidation.NameIsDNSSubdomain
+// ValidateIngressCreate validates Ingresses on create.
+func ValidateIngressCreate(ingress *networking.Ingress, requestGV schema.GroupVersion) field.ErrorList {
+	allErrs := field.ErrorList{}
+	var opts IngressValidationOptions
+	opts = IngressValidationOptions{
+		// TODO: in 1.19, set RequireRegexPath:false
+		// old ingress only has regex-compatible paths
+		requireRegexPath: true}
```
Comment (liggitt) [inline pkg/apis/networking/validation/validation.go]:
```suggestion
		requireRegexPath: true,
	}
```

File: pkg/apis/networking/validation/validation.go [hunk 11]
```diff
@@ -174,15 +182,47 @@ func ValidateIPBlock(ipb *networking.IPBlock, fldPath *field.Path) field.ErrorLi
...
+}
+
+// ValidateIngressUpdate validates ingresses on update.
+func ValidateIngressUpdate(ingress, oldIngress *networking.Ingress, requestGV schema.GroupVersion) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMetaUpdate(&ingress.ObjectMeta, &oldIngress.ObjectMeta, field.NewPath("metadata"))
+	var opts IngressValidationOptions
+	opts = IngressValidationOptions{
+		// TODO: in 1.19, set RequireRegexPath:false
+		// old ingress only has regex-compatible paths
+		requireRegexPath: true}
```
Comment (liggitt) [inline pkg/apis/networking/validation/validation.go]:
```suggestion
		requireRegexPath: allPathsPassRegexValidation(oldIngress),
	}
```

File: pkg/apis/networking/validation/validation.go [hunk 12]
```diff
@@ -174,15 +182,47 @@ func ValidateIPBlock(ipb *networking.IPBlock, fldPath *field.Path) field.ErrorLi
...
 
-// ValidateIngressName validates that the given name can be used as an Ingress name.
-var ValidateIngressName = apimachineryvalidation.NameIsDNSSubdomain
+// ValidateIngressCreate validates Ingresses on create.
+func ValidateIngressCreate(ingress *networking.Ingress, requestGV schema.GroupVersion) field.ErrorList {
+	allErrs := field.ErrorList{}
+	var opts IngressValidationOptions
+	opts = IngressValidationOptions{
+		// TODO: in 1.19, set RequireRegexPath:false
+		// old ingress only has regex-compatible paths
```
Comment (liggitt) [inline pkg/apis/networking/validation/validation.go]: drop the old ingress comment; for the create method, there is no old ingress

File: pkg/apis/networking/validation/validation.go [hunk 13]
```diff
@@ -206,38 +246,32 @@ func validateIngressTLS(spec *networking.IngressSpec, fldPath *field.Path) field
 }
 
 // ValidateIngressSpec tests if required fields in the IngressSpec are set.
-func ValidateIngressSpec(spec *networking.IngressSpec, fldPath *field.Path) field.ErrorList {
+func ValidateIngressSpec(spec *networking.IngressSpec, fldPath *field.Path, opts IngressValidationOptions, requestGV schema.GroupVersion) field.ErrorList {
 	allErrs := field.ErrorList{}
-	// TODO: Is a default backend mandatory?
+	if len(spec.Rules) == 0 && spec.Backend == nil {
+		errMsg := fmt.Sprintf("either `%s` or `rules` must be specified", fldPath.Child("backend"))
```
Comment (liggitt) [inline pkg/apis/networking/validation/validation.go]:
```suggestion
		errMsg := "either `backend` or `rules` must be specified"
```

File: pkg/apis/networking/validation/validation.go [hunk 14]
```diff
@@ -259,45 +293,73 @@ func validateIngressRules(ingressRules []networking.IngressRule, fldPath *field.
...
+	for i, path := range httpIngressRuleValue.Paths {
+		allErrs = append(allErrs, validateHTTPIngressPath(&path, fldPath.Child("paths").Index(i), opts)...)
+	}
+	return allErrs
+}
+
+func validateHTTPIngressPath(path *networking.HTTPIngressPath, fldPath *field.Path, opts IngressValidationOptions) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if path.PathType != nil {
```
Comment (liggitt) [inline pkg/apis/networking/validation/validation.go]: Because of defaulting, PathType should not be able to be nil here. If it is, that's a programming error and we don't want to skip all validation and return valid.

File: pkg/apis/networking/validation/validation.go [hunk 15]
```diff
@@ -259,45 +293,73 @@ func validateIngressRules(ingressRules []networking.IngressRule, fldPath *field.
...
+					// last '/' is a valid regex that matches valid#bar, which isn't
+					// a valid path, because the path terminates at the first ? or
+					// #. A more sophisticated form of validation would detect that
+					// the user is confusing url regexes with path regexes.
+					_, err := regexp.CompilePOSIX(path.Path)
+					if err != nil {
+						allErrs = append(allErrs, field.Invalid(fldPath.Child("path"), path.Path, "must be a valid regex"))
+					}
+				}
+		}
```
Comment (liggitt) [inline pkg/apis/networking/validation/validation.go]: the `/` prefix check and conditional "valid regex" validation must be done for all pathTypes in 1.18 or we'll allow in data that a 1.17 API server would choke on

File: pkg/apis/networking/validation/validation.go [hunk 16]
```diff
@@ -259,45 +293,73 @@ func validateIngressRules(ingressRules []networking.IngressRule, fldPath *field.
...
-			// RFC 3986 section-3.3 cannot appear as a literal in the regex.
-			// Consider the example: http://host/valid?#bar, everything after
-			// the last '/' is a valid regex that matches valid#bar, which
-			// isn't a valid path, because the path terminates at the first ?
-			// or #. A more sophisticated form of validation would detect that
-			// the user is confusing url regexes with path regexes.
-			_, err := regexp.CompilePOSIX(rule.Path)
-			if err != nil {
-				allErrs = append(allErrs, field.Invalid(fldPath.Child("paths").Index(i).Child("path"), rule.Path, "must be a valid regex"))
+			if len(path.Path) > 0 && strings.Contains(path.Path, "//") {
```
Comment (liggitt) [inline pkg/apis/networking/validation/validation.go]: were paths containing `%2f`, `%2F`, `/./`, `/../`, or ending with `/..` or `/.` discussed? Specifically forbidding the double slash without addressing those other patterns seems incomplete if we're trying to avoid obviously problematic paths

File: pkg/apis/networking/validation/validation_test.go [hunk 17]
```diff
@@ -897,102 +900,262 @@ func TestValidateIngress(t *testing.T) {
...
+		tweakIngress       func(ing *networking.Ingress)
+		expectErrsOnFields []string
+	}{
+		"empty path (implementation specific)": {
+			tweakIngress: func(ing *networking.Ingress) {
+				ing.Spec.Rules[0].IngressRuleValue.HTTP.Paths[0].Path = ""
+			},
+			expectErrsOnFields: []string{},
+		},
+		//"valid path (no leading slash)": {
```
Comment (cmluciano) [inline pkg/apis/networking/validation/validation_test.go]: @robscott I forgot about these commented lines here and below

Comment (robscott) [reply pkg/apis/networking/validation/validation_test.go]: no worries, have them covered now, thanks! Will push soon. Do you mind updating this PR to remove the draft status?

Comment (cmluciano) [reply pkg/apis/networking/validation/validation_test.go]: Done

File: pkg/apis/networking/validation/validation.go (hunk 16 above)
Comment (robscott) [reply pkg/apis/networking/validation/validation.go]: Those were not discussed previously as far as I'm aware, but they do all seem reasonable. I've added validation covering each of these use cases. Of course any validation we add here will be incomplete as controllers will likely need to add validation custom to their implementations, but it does seem worthwhile to have a stronger base of validation like these use cases.

File: pkg/apis/networking/validation/validation.go (hunk 15 above)
Comment (robscott) [reply pkg/apis/networking/validation/validation.go]: Updated. To clarify, these are still allowed as part of updates to Ingresses that already have non-regex path types, but otherwise are considered invalid.

File: pkg/apis/networking/types.go (hunk 6 above)
Comment (robscott) [reply pkg/apis/networking/types.go]: Added some clarification to the docs, thanks for catching this!

File: pkg/apis/networking/v1beta1/defaults_test.go [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+				Spec: networkingv1beta1.IngressSpec{
+					Rules: []networkingv1beta1.IngressRule{{
+						IngressRuleValue: networkingv1beta1.IngressRuleValue{
+							HTTP: &networkingv1beta1.HTTPIngressRuleValue{
+								Paths: []networkingv1beta1.HTTPIngressPath{*testCase.expected},
+							},
+						},
+					}},
+				},
+			}
```
[the remark this answers is not in the document]
Comment (robscott) [reply pkg/apis/networking/v1beta1/defaults_test.go]: This has been resolved now.

File: pkg/apis/networking/validation/validation.go [hunk 19]
```diff
@@ -174,15 +184,48 @@ func ValidateIPBlock(ipb *networking.IPBlock, fldPath *field.Path) field.ErrorLi
...
+
+// ValidateIngressUpdate validates ingresses on update.
+func ValidateIngressUpdate(ingress, oldIngress *networking.Ingress, requestGV schema.GroupVersion) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMetaUpdate(&ingress.ObjectMeta, &oldIngress.ObjectMeta, field.NewPath("metadata"))
+	var opts IngressValidationOptions
+	if requestGV == networkingv1beta1.SchemeGroupVersion {
+		opts = IngressValidationOptions{
+			// TODO: in 1.19, set RequireRegexPath:false
+			// old ingress only has regex-compatible paths
+			requireRegexPath: true}
```
[the remark this answers is not in the document]
Comment (robscott) [reply pkg/apis/networking/validation/validation.go]: This has been resolved now.

File: pkg/apis/networking/validation/validation.go [hunk 20]
```diff
@@ -259,45 +296,74 @@ func validateIngressRules(ingressRules []networking.IngressRule, fldPath *field.
...
-		allErrs = append(allErrs, validateIngressBackend(&rule.Backend, fldPath.Child("backend"))...)
 	}
+
+	if path.PathType == nil || *path.PathType == networking.PathTypeImplementationSpecific {
+		if len(path.Path) > 0 {
+			if !opts.requireRegexPath {
+				if !strings.HasPrefix(path.Path, "/") {
+					allErrs = append(allErrs, field.Invalid(fldPath.Child("path"), path.Path, "must be an absolute path"))
+				}
+			} else {
```
[the remark this answers is not in the document]
Comment (robscott) [reply pkg/apis/networking/validation/validation.go]: This has been resolved now.

**Comment by robscott** [comment]
Referencing the enhancement issue for tracking: https://github.com/kubernetes/enhancements/issues/1453

**Comment by robscott** [comment]
@liggitt @cmluciano I think I've resolved all the comments now, please take another look whenever you get a chance.

File: pkg/apis/networking/validation/validation.go (hunk 11 above)
Comment (robscott) [reply pkg/apis/networking/validation/validation.go]: This has been updated now, thanks!

File: pkg/apis/networking/types.go [hunk 21]
```diff
@@ -398,19 +400,51 @@ type HTTPIngressRuleValue struct {
 	// options usable by a loadbalancer, like http keep-alive.
 }
 
-// HTTPIngressPath associates a path regex with a backend. Incoming urls matching
-// the path are forwarded to the backend.
+// PathType represents the type of path referred to by a HTTPIngressPath.
+type PathType string
+
+const (
+	// PathTypeExact matches the URL path exactly.
```
Comment (liggitt) [inline pkg/apis/networking/types.go]: does this include case-sensitivity (e.g. would a path of `/FOO` match a request to `/foo` or vice-versa)?

does this include %-encoding (e.g. would a path of `/foo bar` match a request to `/foo%20bar`?

File: pkg/registry/networking/ingress/strategy_test.go [hunk 22]
```diff
@@ -69,6 +71,11 @@ func newIngress() networking.Ingress {
 
 func TestIngressStrategy(t *testing.T) {
 	ctx := genericapirequest.NewDefaultContext()
+	apiRequest := genericapirequest.RequestInfo{APIGroup: "networking.k8s.io",
+		APIVersion: "v1beta1",
+		Resource:   "ingress",
```
Comment (liggitt) [inline pkg/registry/networking/ingress/strategy_test.go]: should be `ingresses`, right?

File: pkg/apis/networking/validation/validation.go [hunk 23]
```diff
@@ -408,3 +464,15 @@ func validateIngressClassParameters(params *api.TypedLocalObjectReference, fldPa
...
 }
+
+// allPathsPassRegexValidation returns true if the Ingress has paths that all
+// match the Ingress path validation with requireRegexPath enabled.
+func allPathsPassRegexValidation(ingress *networking.Ingress) bool {
+	for _, rule := range ingress.Spec.Rules {
+		errs := validateIngressRuleValue(&rule.IngressRuleValue, field.NewPath(""), IngressValidationOptions{requireRegexPath: true})
+		if len(errs) > 0 {
+			return false
+		}
```
Comment (liggitt) [inline pkg/apis/networking/validation/validation.go]: I'd recommend isolating this to just the path regex check... validateIngressRuleValue does other validation

```suggestion
		if rule.HTTP == nil {
			continue
		}
		for _, path := range rule.HTTP.Paths {
			if len(path) == 0 {
				continue
			}
			if _, err := regexp.CompilePOSIX(rule.Path); err != nil {
				return false
			}
		}
```

File: pkg/apis/networking/validation/validation_test.go [hunk 24]
```diff
@@ -1093,6 +1289,64 @@ func TestValidateIngressUpdate(t *testing.T) {
 			},
 			expectedErrs: field.ErrorList{},
 		},
+		"no regex paths -> no regex paths": {
```
Comment (liggitt) [inline pkg/apis/networking/validation/validation_test.go]: should this be renamed, it actually looks like the `valid regex path -> valid regex path` case

also, do we have a tests that verify we tolerate:
* `invalid regex path -> invalid regex path`
* `invalid regex path -> valid regex path`

File: pkg/apis/networking/validation/validation.go [hunk 25]
```diff
@@ -283,45 +309,75 @@ func validateIngressRules(ingressRules []networking.IngressRule, fldPath *field.
...
+}
+
+func validateHTTPIngressPath(path *networking.HTTPIngressPath, fldPath *field.Path, opts IngressValidationOptions) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if path.PathType == nil {
+		return append(allErrs, field.Required(fldPath.Child("pathType"), "pathType must be specified"))
+	} else if !supportedPathTypes.Has(string(*path.PathType)) {
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("pathType"), *path.PathType, supportedPathTypes.List()))
+	}
```
Comment (liggitt) [inline pkg/apis/networking/validation/validation.go]: nit: prefer making the check that pathType is known/valid part of the switch below:

```go
switch *path.PathType {
case networking.PathTypeExact, networking.PathTypePrefix:
  ...
case networking.PathTypeImplementationSpecific:
  ...
default:
  allErrs = append(allErrs, field.NotSupported(fldPath.Child("pathType"), *path.PathType, supportedPathTypes.List()))
```

File: pkg/apis/networking/validation/validation.go [hunk 26]
```diff
@@ -283,45 +309,75 @@ func validateIngressRules(ingressRules []networking.IngressRule, fldPath *field.
...
+
+	switch *path.PathType {
+	case networking.PathTypeExact, networking.PathTypePrefix:
+		if !strings.HasPrefix(path.Path, "/") {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("path"), path.Path, "must be an absolute path"))
+		}
+		if len(path.Path) > 0 {
+			invalidSequences := []string{"//", "/./", "/../", "%2f", "%2F"}
```
Comment (liggitt) [inline pkg/apis/networking/validation/validation.go]: nit: move invalidSequences and invalidSuffixes to be package vars

File: pkg/apis/networking/validation/validation.go [hunk 27]
```diff
@@ -184,17 +192,26 @@ func ValidateIPBlock(ipb *networking.IPBlock, fldPath *field.Path) field.ErrorLi
...
 }
 
 // ValidateIngressCreate validates Ingresses on create.
-func ValidateIngressCreate(ingress *networking.Ingress) field.ErrorList {
+func ValidateIngressCreate(ingress *networking.Ingress, requestGV schema.GroupVersion) field.ErrorList {
 	allErrs := field.ErrorList{}
-	allErrs = append(allErrs, ValidateIngress(ingress)...)
+	var opts IngressValidationOptions
+	opts = IngressValidationOptions{
+		requireRegexPath: true,
```
Comment (liggitt) [inline pkg/apis/networking/validation/validation.go]: add TODO to remove for 1.19

**Comment by liggitt** [comment]
a few clarifications, tests, and code structure nits, then lgtm

File: pkg/apis/networking/types.go (hunk 21 above)
Comment (robscott) [reply pkg/apis/networking/types.go]: I have updated this to clarify that both exact and prefix matches are case sensitive. I think the %-encoding is a great question, but likely too significant in scope to handle here. A brief survey of implementations shows that behavior here varies already, and since we have not yet had a chance to develop consensus around a specific approach here, I'd rather leave it unspecified until we can. I will work on getting that consensus in place prior to a v1 release.

File: pkg/registry/networking/ingress/strategy_test.go (hunk 22 above)
Comment (robscott) [reply pkg/registry/networking/ingress/strategy_test.go]: Fixed, thanks! Not sure how it worked before?

File: pkg/apis/networking/validation/validation.go (hunk 23 above)
Comment (robscott) [reply pkg/apis/networking/validation/validation.go]: Updated, thanks!

File: pkg/apis/networking/validation/validation_test.go (hunk 24 above)
Comment (robscott) [reply pkg/apis/networking/validation/validation_test.go]: Updated + added additional test cases, thanks!

File: pkg/apis/networking/validation/validation.go (hunk 25 above)
Comment (robscott) [reply pkg/apis/networking/validation/validation.go]: Updated, thanks!

File: pkg/apis/networking/validation/validation.go (hunk 27 above)
Comment (robscott) [reply pkg/apis/networking/validation/validation.go]: Updated, thanks!

File: pkg/apis/networking/validation/validation.go (hunk 26 above)
Comment (robscott) [reply pkg/apis/networking/validation/validation.go]: Updated, thanks!

File: pkg/apis/networking/types.go (hunk 21 above)
Comment (liggitt) [reply pkg/apis/networking/types.go]: sounds good.
